refactor: split test docs and add laravel support (#2021)
api-platform/docs: laravel/testing.md

# Testing the API with Laravel

> For an introduction to testing using API Platform, refer to the [Core Testing Documentation](/core/testing.md), or access the

Reviewer: ```suggestion
For an introduction to testing using API Platform, refer to the [Core Testing Documentation](../core/testing.md), or access the
```

> [Symfony Testing Guide](/symfony/testing.md).

Reviewer: ```suggestion
[Symfony Testing Guide](../symfony/testing.md).
```

API-oriented functional tests, thanks to its integrations with API Platform and [Laravel](https://laravel.com/docs/testing).
- **[PHPUnit](https://phpunit.de)**: A testing framework for writing unit tests for your classes and conducting API-oriented
functional tests, with support for API Platform and [Laravel](https://laravel.com/docs/testing).

> [!TIP]
> Pest is built on top of PHPUnit and introduces additional features along with a syntax inspired by Ruby's RSpec and the
> Jest testing APIs.

## Tests with Pest

> [!TIP]
> Even if you are using Pest, you can also use PHPUnit's assertion API, which can be useful if you're already familiar
> with PHPUnit's assertion API or if you need to perform more complex assertions that aren't available in Pest's expectation API.
> For more information see the [Pest Assertion API](https://pestphp.com/docs/writing-tests#content-assertion-api) documentation.

### Installing Pest

By default, when using Laravel, Pest is pre-configured through the Composer plugin `pestphp/pest-plugin`. You can find this plugin listed in the `allow-plugins` section of your `composer.json` file.

To check the Pest installation, run the following command:

> ```bash

Reviewer: ```suggestion
```console
```

Same everywhere

php artisan test
```

If for some reason, Pest is not installed refer to the [Pest Installation Guide](https://pestphp.com/docs/installation).

> In that case, you can run Pest using:
> ```bash

Reviewer: ```suggestion
In that case, you can run Pest using:

```console
```

./vendor/bin/pest
```

### Writing Functional Tests with Pest

#### Generate the Factory

Using Laravel, you can efficiently test databases by combining seeding with model factories. Model factories allow you
to generate large amounts of test data quickly, while seeding ensures your database is pre-populated with the necessary records.

To create a factory for your model, you can use [Laravel Artisan](https://laravel.com/docs/artisan) command.

> For example, to create a factory for a Book model, run:
> ```bash

Reviewer: ```suggestion
For example, to create a factory for a Book model, run:

```console
```

To do so, learn how to write unit tests with [Pest](https://pestphp.com), [PHPUnit](https://phpunit.de/) and
[Laravel Creating Tests Guide](https://laravel.com/docs/11.x/testing#creating-tests).

## Continuous Integration, Continuous Delivery and Continuous Deployment

Running your test suite in your [CI/CD pipeline](https://en.wikipedia.org/wiki/Continuous_integration) is important to ensure good quality and delivery time.

The API Platform distribution is [shipped with a GitHub Actions workflow](https://github.com/api-platform/api-platform/blob/main/.github/workflows/ci.yml) that builds the Docker images, does a [smoke test](<https://en.wikipedia.org/wiki/Smoke_testing_(software)>) to check that the application's entrypoint is accessible, and runs PHPUnit.

The API Platform Demo [contains a CD workflow](https://github.com/api-platform/demo/tree/main/.github/workflows) that uses [the Helm chart provided with the distribution](../deployment/kubernetes.md) to deploy the app on a Kubernetes cluster.

## Additional and Alternative Testing Tools

You may also be interested in these alternative testing tools (not included in the API Platform distribution):

> - [Hoppscotch](https://docs.hoppscotch.io/features/tests), create functional test for your API

Reviewer: Hoppstoch is also on the next line

Author: This was already the case but I deleted the 2nd one.

- [Hoppscotch](https://docs.hoppscotch.io/documentation/features/rest-api-testing/), create functional test for your API
  Platform project using a nice UI, benefit from its Swagger integration and run tests in the CI using [the command-line tool](https://docs.hoppscotch.io/cli);
- [Behat](https://behat.org), a [behavior-driven development (BDD)](https://en.wikipedia.org/wiki/Behavior-driven_development) framework to write the API specification as user
  stories and in natural language then execute these scenarios against the application to validate its behavior;
- [Blackfire Player](https://blackfire.io/player), a nice DSL to crawl HTTP services, assert responses, and extract data
  from HTML/XML/JSON responses;

> - [PHP Matcher](https://github.com/coduo/php-matcher), the Swiss Army knife of JSON document testing.

Reviewer: You can remove these two ones imho

- [Playwright](https://playwright.dev) is recommended if you use have PWA/JavaScript-heavy app.

## Testing Utilities for Laravel

### API Test Assertions with Laravel

> In addition to [the built-in ones](https://phpunit.readthedocs.io/en/11.4/assertions.html), API Platform provides

Reviewer: ```suggestion
In addition to [the built-in ones](https://phpunit.readthedocs.io/en/main/assertions.html), API Platform provides
```
